### PR TITLE
Tests for _Doc, _El, _Obj, and _Str

### PR DESCRIPTION
**Pull request by adewaskar**
Written some mocha tests for _Str and _Obj utility functions.

### CODE REVIEW

File: src/fe/implicit/_Obj.js [hunk 1]
```diff
@@ -1,46 +1,140 @@
...
+ *
+ * @returns {Object} New object
+ */
 export const create = (obj, props) => Object.create(obj, props);
+
+/**
+ * Check if the object has any key value pairs or is empty
+ * @param {Object} o
+ *
+ * @returns {boolean} Returns true or false if the object was empty or not
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: The description for this feels pretty redundant.

File: src/fe/implicit/_Obj.js [hunk 2]
```diff
@@ -1,46 +1,140 @@
+import * as _ from './_';
+import * as _Arr from './_Arr';
+/**
+ * Get keys of an object in an array.
+ * @param {Object} o
+ *
+ * @returns {Array} Array of keys
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: Description seems redundant.

File: src/fe/implicit/_Obj.js (hunk 1 above)
Comment (adewaskar) [reply src/fe/implicit/_Obj.js]: seems to me too. ok repairing it.

File: src/fe/implicit/_Obj.js [hunk 3]
```diff
@@ -1,46 +1,140 @@
...
+ *
+ * @returns {*}
+ */
 export const getOwnProp = _.curry2((o, prop) => hasOwnProp(o, prop) && o[prop]);
 
+/**
+ * Set the property of the given object with the given value
+ * @param {*} subject
+ * @param {Object} o
+ * @param {string} k
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]:
```suggestion
 * @param {string} key
```

File: src/fe/implicit/_Obj.js [hunk 4]
```diff
@@ -1,46 +1,140 @@
...
+ * @param {string} key
+ *
+ * @returns {Object}
+ */
 export const deleteProp = _.curry2((o, key) => {
   delete o[key];
   return o;
 });
 
+// remaining
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]:
```suggestion
/**
 * Loops through the object.
 * @param {Object} o
 * @param {function (value: *, key: *, o: Object): void}
 * 
 * @returns {Object}
 */
```

File: src/fe/implicit/_Obj.js [hunk 5]
```diff
@@ -1,46 +1,140 @@
...
 export const loop = _.curry2((o, iteratee) => {
   _Arr.loop(keys(o), key => iteratee(o[key], key, o));
   return o;
 });
 
-// {a: 2, b: 3} → map(x => 2*x) → {a: 4, b: 6}
+/**
+ * Map the properties of the object to the given condition recieved from the return value of the callback function
+ * @param {Object} o
+ * @param {string} key
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]:
```suggestion
 * @param {function (value: *, key: *, o: Object): *}
```

File: src/fe/implicit/_Obj.js [hunk 6]
```diff
@@ -1,46 +1,140 @@
...
 
+// remaining
 export const loop = _.curry2((o, iteratee) => {
   _Arr.loop(keys(o), key => iteratee(o[key], key, o));
   return o;
 });
 
-// {a: 2, b: 3} → map(x => 2*x) → {a: 4, b: 6}
+/**
+ * Map the properties of the object to the given condition recieved from the return value of the callback function
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]:
```suggestion
 * Loop through the object and map properties
```

File: src/fe/implicit/_Obj.js [hunk 7]
```diff
@@ -1,46 +1,140 @@
...
+ * @returns {Object}
+ */
 export const map = _.curry2((o, iteratee) =>
   _Arr.reduce(
     keys(o),
     (obj, key) => setProp(obj, key, iteratee(o[key], key, o)),
     {}
   )
 );
+// {a: 2, b: 3} → map(x => 2*x) → {a: 4, b: 6}
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: Add this as @example
https://jsdoc.app/tags-example.html

File: src/fe/implicit/_Obj.js [hunk 8]
```diff
@@ -1,46 +1,140 @@
...
 
+/**
+ * Loops on the object and apply a reducer function on it
+ * @param {Object} o
+ * @param {function} reducer
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: Add params and return value

Comment (umanghome) [inline src/fe/implicit/_Obj.js]:
```suggestion
 * @param {function (accumulator: *, currentValue: *): *} reducer
```

File: src/fe/implicit/_Obj.js [hunk 9]
```diff
@@ -49,8 +143,20 @@ export const reduce = _.curry3((o, reducer, initialValue) =>
...
+ *
+ * @returns {string}
+ */
 export const stringify = JSON.stringify;
 
+/**
+ * Parse a string into JSON
+ * @param {string} string
+ *
+ * @returns {string}
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]:
```suggestion
 * @returns {Object}
```

File: src/fe/implicit/_Obj.js [hunk 10]
```diff
@@ -49,8 +143,20 @@ export const reduce = _.curry3((o, reducer, initialValue) =>
...
+/**
+ * Stringify an object
+ * @param {Object} o
+ *
+ * @returns {string}
+ */
 export const stringify = JSON.stringify;
 
+/**
+ * Parse a string into JSON
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: Document exception as well.
https://jsdoc.app/tags-throws.html

File: src/fe/implicit/_Obj.js [hunk 11]
```diff
@@ -59,13 +165,31 @@ export const parse = string => {
   }
 };
 
+/**
+ * Clone an object from the previous object
+ * @param {Obect} o
+ *
+ * @returns {Object}
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: This also throws an exception due to usage of `parse`

File: src/fe/implicit/_Obj.js [hunk 12]
```diff
@@ -59,13 +165,31 @@ export const parse = string => {
...
+ */
 export const clone = o => parse(stringify(o));
 
+/**
+ * Extend an object with the properties from the given object
+ * @param {string} string
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: Fix args

File: src/fe/implicit/_Obj.js [hunk 13]
```diff
@@ -97,6 +221,12 @@ export const unflatten = o => {
   return result;
 };
 
+/**
+ * Flattens the object by turning nested object into object with delimitters in the keys
+ * @param {Object} o
+ *
```
Comment (umanghome) [inline src/fe/implicit/_Obj.js]: Add prefix arg with default value

File: src/fe/implicit/_Str.js [hunk 14]
```diff
@@ -1,11 +1,51 @@
+import * as _ from './_';
 // https://vocajs.com/
 const proto = _.prototypeOf(String);
 const protoSlice = proto.slice;
 
+/**
+ * Pass a string and add padding to its ends
```
Comment (umanghome) [inline src/fe/implicit/_Str.js]:
```suggestion
 * Pass a string and add padding to its beginning and end
```

File: tests/_Str.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  deepEqual: deep,
+  notDeepEqual: notDeep,
+  equal,
+} = assert;
+
+describe('_Str', () => {
+  describe('pad', () => {
+    it('Check padding on both sides', () => {
+      const arr = _Str.pad('test').split('');
+      const hasPadding = arr[0] == ' ' && arr[arr.length - 1] == ' ';
```
Comment (umanghome) [inline tests/_Str.js]:
```suggestion
      const hasPadding = arr[0] === ' ' && arr[arr.length - 1] === ' ';
```

File: tests/_Str.js [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+    it('Check if a string does not contain a substring', () => {
+      const str = 'hello world';
+      const substr = 'world1';
+      const contains = _Str.contains(str, substr);
+      isFalse(contains);
+    });
+  });
+
+  describe('slice', () => {
```
Comment (umanghome) [inline tests/_Str.js]: Add tests where `from` is negative and another where `to` is greater than the length

File: tests/_Str.js [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    it('Check if a string is sliced for given indexes', () => {
+      const str = 'hello testing';
+      const from = 2;
+      const to = 5;
+      const sliced = _Str.slice(str, from, to);
+      equal(sliced, 'llo');
+    });
+  });
+
+  describe('sliceFrom', () => {
```
Comment (umanghome) [inline tests/_Str.js]: Add tests where `from` is negative

File: tests/_Str.js [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    it('Check if a string is sliced from a given index', () => {
+      const str = 'hello testing';
+      const from = 2;
+      const sliced = _Str.sliceFrom(str, from);
+      equal(sliced, 'llo testing');
+    });
+  });
+
+  describe('startsWith', () => {
```
Comment (umanghome) [inline tests/_Str.js]: add negative case

File: tests/_Obj.js [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+
+describe('_Obj', () => {
+  describe('keys', () => {
+    it('Get keys of the object', () => {
+      const obj = {
+        a: 1,
+        b: 2,
+      };
+      const keys = _Obj.keys(obj);
+      const isCorrect = keys[0] == 'a' && keys[1] == 'b';
```
Comment (umanghome) [inline tests/_Obj.js]:
```suggestion
      const expected = ['a', 'b'];
```

File: tests/_Obj.js [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+      isTrue(isCorrect && Array.isArray(keys));
```
Comment (umanghome) [inline tests/_Obj.js]:
```suggestion
      deep(keys, expected);
```

File: tests/_Obj.js [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+    });
+  });
+
+  describe('create', () => {
```
Comment (umanghome) [inline tests/_Obj.js]: TODO

File: tests/_Obj.js [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+    });
+
+    it('Check if the object is not empty', () => {
+      const obj = { test: 1 };
+      const isEmpty = _Obj.isEmpty(obj);
+      isFalse(isEmpty);
+    });
+  });
+
+  describe('hasProp', () => {
```
Comment (umanghome) [inline tests/_Obj.js]: Check negative condition as well

File: tests/_Obj.js [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+  });
+
+  describe('loop', () => {
+    it('Check if it loops properly on the object', () => {
+      const obj = { a: 1, b: 1, c: 4 };
+      const arr = [];
+      _Obj.loop(obj, k => {
+        arr.push(k);
+      });
+      const isLoopWorking = arr[0] == 1 && arr[1] == 1 && arr[2] == 4;
```
Comment (umanghome) [inline tests/_Obj.js]: Convert this to `expected` and use `deep` to check

File: tests/_Obj.js [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+      isTrue(isLoopWorking);
+    });
+  });
+
+  describe('setTruthyProp', () => {
```
Comment (umanghome) [inline tests/_Obj.js]: You're calling `setProp` here. Call `setTruthyProp` instead

File: tests/_Obj.js [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+    it('Check if it sets the truthy property on the object', () => {
+      const obj = { a: 1 };
+      _Obj.setProp(obj, 'b', 2);
+      isTrue(obj.b == 2);
+    });
+
+    it('Check if it does not set untruthy property on the object', () => {
+      const obj = { a: 1 };
+      _Obj.setProp(obj, 'b', false);
```
Comment (umanghome) [inline tests/_Obj.js]: You're calling `setProp` here. Call `setTruthyProp` instead

File: tests/_Obj.js [hunk 26]
```diff
@@ -0,0 +1,222 @@
...
+      });
+      const isLoopWorking = arr[0] == 1 && arr[1] == 1 && arr[2] == 4;
+      isTrue(isLoopWorking);
+    });
+  });
+
+  describe('setTruthyProp', () => {
+    it('Check if it sets the truthy property on the object', () => {
+      const obj = { a: 1 };
+      _Obj.setProp(obj, 'b', 2);
```
Comment (umanghome) [inline tests/_Obj.js]: You're calling `setProp` here. Call `setTruthyProp` instead

File: tests/_Obj.js [hunk 27]
```diff
@@ -0,0 +1,222 @@
...
+      _Obj.setProp(obj, 'b', false);
+      isFalse(obj.b);
+    });
+  });
+
+  describe('deleteProp', () => {
+    it('Check if it deleted a property on the object', () => {
+      const obj = { a: 1, b: 1 };
+      _Obj.deleteProp(obj, 'b');
+      isTrue(obj.b === undefined);
```
Comment (umanghome) [inline tests/_Obj.js]:
```suggestion
      equal(typeof obj.b, 'undefined');
```

File: tests/_Obj.js [hunk 28]
```diff
@@ -0,0 +1,222 @@
...
+
+    it('Check if it does not set untruthy property on the object', () => {
+      const obj = { a: 1 };
+      _Obj.setProp(obj, 'b', false);
+      isFalse(obj.b);
+    });
+  });
+
+  describe('deleteProp', () => {
+    it('Check if it deleted a property on the object', () => {
```
Comment (umanghome) [inline tests/_Obj.js]: Also add another `it` which checks that `obj.a` is not deleted.

File: tests/_Obj.js [hunk 29]
```diff
@@ -0,0 +1,222 @@
...
+      _Obj.deleteProp(obj, 'b');
+      isTrue(obj.b === undefined);
+    });
+  });
+
+  describe('map', () => {
+    it('Check if it correctly maps the object and returns an object', () => {
+      const obj = { a: 1, b: 1 };
+      const mappedObj = _Obj.map(obj, k => k * 3);
+      isTrue(mappedObj.a == 3 && mappedObj.b == 3);
```
Comment (umanghome) [inline tests/_Obj.js]: Use `expected` and `deep`

File: tests/_Obj.js [hunk 30]
```diff
@@ -0,0 +1,222 @@
...
+
+  describe('stringify', () => {
+    it('Check if it correctly stringifies an object', () => {
+      const obj = _Obj.parse(_Obj.stringify({ a: 1 }));
+      isTrue(obj.a == 1);
+    });
+  });
+
+  describe('parse', () => {
+    it('Check if it correctly parses an stringified object', () => {
```
Comment (umanghome) [inline tests/_Obj.js]: Pass invalid JSON and check if an exception is thrown.

File: tests/_Obj.js [hunk 31]
```diff
@@ -0,0 +1,222 @@
...
+      const obj = {};
+      const reduced = _Obj.reduce(obj, reducer, 3);
+      isTrue(reduced === 3);
+    });
+  });
+
+  describe('stringify', () => {
+    it('Check if it correctly stringifies an object', () => {
+      const obj = _Obj.parse(_Obj.stringify({ a: 1 }));
+      isTrue(obj.a == 1);
```
Comment (umanghome) [inline tests/_Obj.js]: Use `deep` for checking this.

File: tests/_Obj.js [hunk 32]
```diff
@@ -0,0 +1,222 @@
...
+    });
+  });
+
+  describe('parse', () => {
+    it('Check if it correctly parses an stringified object', () => {
+      const obj = _Obj.parse(_Obj.stringify({ a: 1 }));
+      isTrue(obj.a == 1);
```
Comment (umanghome) [inline tests/_Obj.js]: Use `deep` for checking this.

File: tests/_Obj.js [hunk 33]
```diff
@@ -0,0 +1,222 @@
...
+    });
+  });
+
+  describe('clone', () => {
+    it('Check if it correctly clones a stringified object', () => {
+      const obj = { a: 1 };
+      const obj2 = _Obj.clone(obj);
+      isTrue(obj !== obj2 && obj.a == obj2.a);
```
Comment (umanghome) [inline tests/_Obj.js]: Use `deep`

Comment (umanghome) [inline tests/_Obj.js]: And check for thrown exceptions

File: tests/_Obj.js [hunk 34]
```diff
@@ -0,0 +1,222 @@
...
+    });
+  });
+
+  describe('extend', () => {
+    it('Check if it correctly extends an object', () => {
+      const obj = { a: 1 };
+      const obj2 = _Obj.extend(obj, { b: 1 });
+      isTrue(obj2.a == 1 && obj2.b == 1);
```
Comment (umanghome) [inline tests/_Obj.js]: Use `deep`

File: tests/_Obj.js [hunk 35]
```diff
@@ -0,0 +1,222 @@
...
+      isFalse(hasOwnProp);
+    });
+  });
+
+  describe('setPropOf', () => {
+    it('Check if it correctly sets the prop', () => {
+      const obj = { a: 1 };
+      const obj2 = { b: 1 };
+      _Obj.setPropOf(obj, obj2, 'c');
+      isTrue(obj2.c.a === 1);
```
Comment (umanghome) [inline tests/_Obj.js]: `deep`

File: tests/_Obj.js [hunk 36]
```diff
@@ -0,0 +1,222 @@
...
+    });
+  });
+
+  describe('flattens', () => {
+    it('Check if it correctly flattens an object', () => {
+      const obj = { a: { b: 1 } };
+      const obj2 = _Obj.flatten(obj);
+      isTrue(obj2['a.b'] == 1);
```
Comment (umanghome) [inline tests/_Obj.js]: Use `deep`

**Comment by umanghome** [comment]
Can you please make individual PRs for individual modules? This PR should have only contained `_Str` and `_Obj`

File: package.json [hunk 37]
```diff
@@ -7,7 +7,7 @@
     "coverage": "nyc --require @babel/register npm run test",
     "lint-documentation": "documentation lint src --shallow",
     "start": "rollup -cw",
-    "test": "mocha --require @babel/register \"tests/**/*.js\"",
+    "test": "mocha --require @babel/register \"tests/**/*.js\" -r jsdom-global/register",
```
Comment (umanghome) [inline package.json]:
```suggestion
    "test": "mocha --require @babel/register --require jsdom-global/register \"tests/**/*.js\"",
```

File: play/implicit/src/index.js [hunk 38]
```diff
@@ -1,7 +1,12 @@
...
+  _Obj,
+  _Doc,
+  _El,
 };
 
 Object.keys(EXPOSED_MODULES).forEach(key => {
   global[key] = EXPOSED_MODULES[key];
 });
+
+console.log(_Doc);
```
Comment (umanghome) [inline play/implicit/src/index.js]:
```suggestion
```

File: src/fe/implicit/_Doc.js [hunk 39]
```diff
@@ -11,16 +14,38 @@ export const querySelectorAll = _Func.bind('querySelectorAll', document);
...
 export const EventConstructor = global.Event;
+import * as _El from './_El';
+import * as _Obj from './_Obj';
+var link;
+
 export const isEvent = x => _.is(x, EventConstructor);
+
+/**
+ * Resolves given string to an element.
+ * @param {*} el
```
Comment (umanghome) [inline src/fe/implicit/_Doc.js]:
```suggestion
 * @param {string|Element} el
```

File: src/fe/implicit/_El.js [hunk 40]
```diff
@@ -1,6 +1,20 @@
 export const ElementConstructor = global.Element;
+import * as _ from './_';
 
+/**
+ * Creates an element using a tag name.
+ * @param {string} tagName
+ *
+ * @returns {Object}
```
Comment (umanghome) [inline src/fe/implicit/_El.js]:
```suggestion
 * @returns {Element}
```

File: src/fe/implicit/_El.js [hunk 41]
```diff
@@ -1,6 +1,20 @@
...
+ */
 export const create = tagName => document.createElement(tagName || 'div');
+
+/**
+ * Gets the parent element of the given element.
+ * @param {Object} element
```
Comment (umanghome) [inline src/fe/implicit/_El.js]:
```suggestion
 * @param {Element} element
```

File: src/fe/implicit/_El.js [hunk 42]
```diff
@@ -1,6 +1,20 @@
...
+ *
+ * @returns {Object}
```
Comment (umanghome) [inline src/fe/implicit/_El.js]:
```suggestion
 * @returns {Element}
```

File: src/fe/implicit/_Doc.js [hunk 43]
```diff
@@ -83,6 +130,12 @@ export function form2obj(form) {
   );
 }
 
+/**
+ * Prevents default event from firing
+ * @param {Object} e
```
Comment (umanghome) [inline src/fe/implicit/_Doc.js]:
```suggestion
 * @param {Event} e
```

**Comment by adewaskar** [comment]
yay!
